File: Array/Binary Search/binarySearchVersion2.java
package com.algorithm.array;
import org.junit.Test;

/**
 * TODO: Binary Search Version2
 *
 * @author Yaozheng Wang
 * @date 2021-12-21
 */
public class binarySearchVersion2 {
    public int findPosition (int[] numbs, int target) {
        // Consider the list is null or length is 0
        if (numbs == null || numbs.length == 0) {
            return -1;
        }

        // Initialize the left index and right index
        // to be length minus 1 because of [left, right]
        int left = 0, right = numbs.length - 1;

        // Use while loop to consider three situations
        // The left and right index can be the same due to [left, right]
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (numbs[mid] == target) {
                return mid;
            } else if (numbs[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        // No match number
        return -1;
    }

    @Test
    public static void main(String[] args) {
        binarySearchVersion2 question = new binarySearchVersion2();
        int[] numbs = new int[]{1, 2, 3, 4, 5};
        System.out.println(question.findPosition(numbs, 3));
        System.out.println(question.findPosition(numbs, 9));
    }
}

File: Array/Binary Search/binary_search_version2.py
from typing import List


class Solution:
    def binary_search(self, nums: List[int], target: int) -> int:
        """
        The methods binary search version2 that [left, right)
        :param nums: integer list
        :param target: the target number
        :return: -1 if no result, otherwise
            the target number index in the list
        """
        # Initialize the left index, right index.
        # The right index can be the list length dut to [left, right)
        left, right = 0, len(nums)
        # Use while loop to do binary search, notice that left cannot equal to right
        while left < right:
            middle = left + (right - left) // 2
            # Consider three situations
            if nums[middle] < target:
                left = middle + 1
            elif nums[middle] > target:  # right index can be middle index because of [left, right)
                right = middle
            else:
                return middle
        # The situation that no matching number
        return -1


def main():
    question1 = Solution()
    solution1 = question1.binary_search([1, 3, 4, 6, 7], 4)
    print(solution1)

    question2 = Solution()
    solution2 = question2.binary_search([1, 3, 4, 6, 7], 10)
    print(solution2)


if __name__ == "__main__":
    main()
